### DISCUSSION TITLE
Why a std::array of MultiFabs instead of a single MultiFab with ncomp=2 in Electrostatic PIC tutorial?

### DISCUSSION THREAD

**Discussion by n8xm** [AMReX-Codes/amrex#2488] [Q&A] [answered]
In the [Electrostatic PIC tutorial](https://amrex-codes.github.io/amrex/tutorials_html/Particles_Tutorial.html#electrostaticpic) ( authored by @atmyers ), we set up `eField` (which I am pretty sure is the electric field) as follows:
1. Initialize `eField` on [line 104](https://github.com/AMReX-Codes/amrex-tutorials/blob/188babb427d8e0518a4ef74244aeab4e6e6036d8/ExampleCodes/Particles/ElectrostaticPIC/src/main.cpp#L104),  as a `Vector` of `std::array` of `MultiFab` pointers, where the `std::array` is of length `AMREX_SPACEDIM`. 
2. On [lines 114 to 117](https://github.com/AMReX-Codes/amrex-tutorials/blob/188babb427d8e0518a4ef74244aeab4e6e6036d8/ExampleCodes/Particles/ElectrostaticPIC/src/main.cpp#L114-L117), for each spatial dimension, create a new `MultiFab` with the `ncomp` argument set to `1` 

Presumably, the reason why we effectively have an have an array of `MultiFab`s of length `AMREX_SPACEDIM` is because the electric field is a vector field, unlike `phi` (electric potential?) and `rhs` (RHS of the Poisson problem?) which appear to be scalar fields. 

What I'm wondering is: If `eField` is a vector field (which by definition requires one component per spatial dimension), why create multiple `MultiFab`s, each with `ncomp=1`, as opposed to a single `MultiFab` with `ncomp=AMREX_SPACEDIM`? Is it because the [compute_E_nodal function](https://github.com/AMReX-Codes/amrex-tutorials/blob/188babb427d8e0518a4ef74244aeab4e6e6036d8/ExampleCodes/Particles/ElectrostaticPIC/src/field_solver/FieldSolver_K.H#L79) assumes separate arrays for each spatial dimension? Or is there some other advantage?

**Comment by atmyers** [comment, marked as answer]
In that particular example, the electric field could be stored in a multi-component Multifab as you suggest, and the code would likely be a cleaner, because all the components of E have the same nodal centering. However, if you have a case where the different components might have different staggering, like on a [Yee grid](https://en.wikipedia.org/wiki/Finite-difference_time-domain_method#/media/File:Yee_cell.png), then you need to store then as separate Multifabs. I think I was probably starting from a more general code that had (or could potentially have based on runtime options) staggered E fields.
